lib/ngx/resolver.lua

```diff
@@ -0,0 +1,85 @@
+local base = require "resty.core.base"
+local get_request = base.get_request
+local ffi = require "ffi"
+local C = ffi.C
+local ffi_new = ffi.new
+local ffi_str = ffi.string
+local ffi_gc = ffi.gc
+local FFI_OK = base.FFI_OK
+local FFI_ERROR = base.FFI_ERROR
+local FFI_DONE = base.FFI_DONE
+local co_yield = coroutine._yield
+
+local BUF_SIZE = 256
+local get_string_buf = base.get_string_buf
+local get_size_ptr = base.get_size_ptr
+
+base.allows_subsystem("http")
+
+ffi.cdef [[
+typedef intptr_t                         ngx_int_t;
+typedef unsigned char                    u_char;
+typedef struct ngx_http_lua_co_ctx_t    *curcoctx_ptr;
+typedef struct ngx_http_resolver_ctx_t  *rctx_ptr;
+
+typedef struct {
+    ngx_http_request_t              *request;
+    u_char                          *buf;
+    size_t                          *buf_size;
+    curcoctx_ptr                     curr_co_ctx;
+    rctx_ptr                         rctx;
+    ngx_int_t                        rc;
+    unsigned                         ipv4:1;
+    unsigned                         ipv6:1;
+} ngx_http_lua_resolver_ctx_t;
+
+int ngx_http_lua_ffi_resolve(ngx_http_lua_resolver_ctx_t *ctx, const char *hostname);
+
+void ngx_http_lua_ffi_resolver_destroy(ngx_http_lua_resolver_ctx_t *ctx);
+]]
+
+local _M = { version = base.version }
+
+local mt = {
+    __gc = C.ngx_http_lua_ffi_resolver_destroy
+}
+
+local Ctx = ffi.metatype("ngx_http_lua_resolver_ctx_t", mt)
+
+function _M.resolve(hostname, ipv4, ipv6)
+    assert(type(hostname) == "string", "hostname must be string")
+    assert(ipv4 == nil or type(ipv4) == "boolean", "ipv4 must be boolean or nil")
+    assert(ipv6 == nil or type(ipv6) == "boolean", "ipv6 must be boolean or nil")
+
+    local buf = get_string_buf(BUF_SIZE)
+    local buf_size = get_size_ptr()
+    buf_size[0] = BUF_SIZE
+
+    local ctx = Ctx({get_request(), buf, buf_size})
+
+    ctx.ipv4 = ipv4 == nil and 1 or ipv4
+    ctx.ipv6 = ipv6 == nil and 0 or ipv6
+
+    local rc = C.ngx_http_lua_ffi_resolve(ctx, hostname)
```

Comment:
You know that `ctx` is managed by Lua GC while your `ctx` may get collected as soon as this Lua function returns (or yields). I see you register `ctx` to `rctx` which will surely outlive this Lua function call once it yields.

Reply:
I really believed that coroutine function locals must not be collected by GC, until function returns (yield is not a case when local might be get freed by GC).

I spent several hours for reading manual and googling. Unfortunately, I did not find anything that clearly states what should be during a yield with locals.

Based on following I wrote several tests.
Lua: test suites https://www.lua.org/tests/ (for v5.1)
Lua Traverse http://code.matthewwild.co.uk/luatraverse/

```
local object_was_finalized

local function finalizer()
    object_was_finalized = true
end

local function func()
    local object = newproxy(true)
    getmetatable(object).__gc = finalizer
    local free = coroutine.yield()
    if free then
        object = nil
    end
    coroutine.yield()
end

-- Test #1
object_was_finalized = false
local t = coroutine.create(func)
coroutine.resume(t)
collectgarbage()
assert(object_was_finalized == false)
coroutine.resume(t, true)
collectgarbage()
assert(object_was_finalized == true)

-- Test #2 
object_was_finalized = false
t = coroutine.create(func)
coroutine.resume(t)
collectgarbage()
assert(object_was_finalized == false)
coroutine.resume(t, false)
collectgarbage()
assert(object_was_finalized == false)
coroutine.resume(t)
collectgarbage()
assert(object_was_finalized == true)
```
... and still not sure, if local variable can be GC-ed on yield.

Could you please point me to manual/link which covers this (GC on yield).

Reply:
Additionally, I created a test with explicit lua_gc calls (from C-land) just before return into yielded-lua-thread and right after the return from resumed thread.
The test confirmed assertion that Ctx can get freed only after return from coroutine (not in the middle on yield).

Please, give me a direction/additional info on this.